src/web/TextInput.tsx

```diff
@@ -43,6 +43,72 @@ export interface TextInputContext {
     focusArbitrator?: FocusArbitratorProvider;
 }
 
+class TextInputPlaceholderSupport {
+    private _internalId = TextInputPlaceholderSupport.sequenceId();
+    private _styleElement: HTMLStyleElement | null = null;
+
+    private static sequenceId = (() => {
+        let id = 0;
+        return () => ++id;
+    })();
+
+    className(): string {
+        return `reactxp-text-input-${this._internalId}`;
+    }
+
+    tearDown() {
+        const styleElement = this._styleElement;
+        if (styleElement && styleElement.parentNode) {
+            styleElement.parentNode.removeChild(styleElement);
+        }
+    }
+
+    private _setupStyleElement(): HTMLStyleElement {
+        let style = this._styleElement;
+
+        if (!style) {
+            style = document.createElement('style');
+            style.type = 'text/css';
+            this._styleElement = style;
+        }
+
+        if (!style.parentNode) {
+            document.head.appendChild(style);
+        }
+
+        return style;
+    }
+
+    setColor(placeholderColor: string | undefined) {
+        if (typeof document === undefined) {
```

Comment:
Some of existing code does explicit check for `document`. I suppose this guards ReactXP from crashing on non-DOM test environments.